packages/@uppy/provider-views/src/View.ts

```diff
@@ -1,11 +1,91 @@
+import type {
+  UnknownProviderPlugin,
+  UnknownSearchProviderPlugin,
+} from '@uppy/core/lib/Uppy'
+import type { Body, Meta } from '@uppy/utils/lib/UppyFile'
+import type { CompanionFile } from '@uppy/utils/lib/CompanionFile'
 import getFileType from '@uppy/utils/lib/getFileType'
 import isPreviewSupported from '@uppy/utils/lib/isPreviewSupported'
 import remoteFileObjToLocal from '@uppy/utils/lib/remoteFileObjToLocal'
 
-export default class View {
-  constructor (plugin, opts) {
+type TagFile<M extends Meta> = {
+  id: string
+  source: string
+  data: Blob
+  name: string
+  type: string
+  isRemote: boolean
+  preview?: string
+  meta: {
+    authorName?: string
+    authorUrl?: string
+    relativePath?: string | null
+    absolutePath?: string
+  } & M
+  remote: {
+    companionUrl: string
+    url: string
+    body: {
+      fileId: string
+    }
+    providerName: string
+    provider: string
+    requestClientId: string
+  }
+}
+
+type PluginType = 'Provider' | 'SearchProvider'
+
+// Conditional type for selecting the plugin
+type SelectedPlugin<M extends Meta, B extends Body, T extends PluginType> =
+  T extends 'Provider' ? UnknownProviderPlugin<M, B>
+  : T extends 'SearchProvider' ? UnknownSearchProviderPlugin<M, B>
+  : never
+
+// Conditional type for selecting the provider from the selected plugin
+type SelectedProvider<
+  M extends Meta,
+  B extends Body,
+  T extends PluginType,
+> = SelectedPlugin<M, B, T>['provider']
+
+export interface ViewOptions<
+  M extends Meta,
+  B extends Body,
+  T extends PluginType,
+> {
+  provider: SelectedProvider<M, B, T>
+  viewType?: string
+  showTitles?: boolean
+  showFilter?: boolean
+  showBreadcrumbs?: boolean
+  loadAllFiles?: boolean
+}
+
+export default class View<
+  M extends Meta,
+  B extends Body,
+  T extends PluginType,
+  O extends ViewOptions<M, B, T>,
+> {
+  plugin: SelectedPlugin<M, B, T>
+
+  provider: SelectedProvider<M, B, T>
+
+  isHandlingScroll: boolean
+
+  requestClientId: string
+
+  isShiftKeyPressed: boolean
+
+  lastCheckbox: CompanionFile | undefined
+
+  opts: O
+
+  constructor(plugin: SelectedPlugin<M, B, T>, opts: O) {
     this.plugin = plugin
     this.provider = opts.provider
+    this.opts = opts
```

Comment:
Let's not add more public properties
```suggestion
```

Reply:
This is done similarly to `BasePlugin`, options are set here and only merged in the parent classes. I think it makes more sense like this for now until we remove the class.